### PR TITLE
OpenVDBLink -- link between Mathematica and OpenVDB

### PR DESCRIPTION
**Pull request by ghurstunither**
## Dependencies

Current dependencies are just OpenVDB and Mathematica itself.

We can compile the link in a Wolfram notebook.

------------

## Compilation

Here's an small function that will load the library:

    LoadVDB[] := 
      Block[{vdbpath},
        
        vdbpath = FileNameJoin[{"Your_OpenVDB_Location", "openvdb", "openvdb_wolfram"}];
        
        $ContextPath = DeleteDuplicates[Join[$ContextPath, {"OpenVDBLink`"}]];
        $Path = DeleteDuplicates[Join[$Path, {vdbpath}]];
        
        <<OpenVDBLink`
      ]

Once `LoadVDB[]` is ran we can compile. Any specific compilation options are preset in `BuildSettings.m`. Once set run

    OpenVDBLink`Developer`Recompile[]

This will take a few minutes to run. This function takes an optional boolean argument, and when set to `True` the compilation command will be printed in the notebook as well as any warnings or errors generated during compilation.

------------

## Testing

Once compiled, you can run the unit tests to make sure everything built correctly.

    testresults = OpenVDBLink`Developer`TestOpenVDBLink[];

The test results are returned in nested [`Association`](https://reference.wolfram.com/language/ref/Association.html) (think Python dictionary) form. The results use Wolfram's native [`TestReportObject`](https://reference.wolfram.com/language/ref/TestReportObject.html), and can be inspected, etc. Global stats are also returned, 1392 tests for example.

Screen Shot 2022-04-09 at 2 09 51 PM

------------

### List of functionality ported over (so far)

    ??OpenVDBLink`*

Screen Shot 2022-04-09 at 3 24 26 PM

### Supported grid types

Here, `"Scalar"` is a synonym for `"Float"` and `"Vector"` for `"Vec3S"`:

    OpenVDBGridTypes[]

>     {"Scalar", "Vector", "Double", "Float", "Byte", "Int32", "Int64", "UInt32", 
>      "Vec2D", "Vec2I", "Vec2S", "Vec3D", "Vec3I", "Vec3S", "Boolean", "Mask"}

An empty grid with voxel spacing of `1.0` for each type:

    OpenVDBCreateGrid[1.0, #] & /@ OpenVDBGridTypes[]

Screen Shot 2022-04-09 at 3 05 22 PM

### Simple example

Get a triangle mesh of a Triceratops through [`ExampleData`](https://reference.wolfram.com/language/ref/ExampleData.html):

    dino = ExampleData[{"Geometry3D", "Triceratops"}, "MeshRegion"]

Screen Shot 2022-04-09 at 2 27 30 PM

Create a grid with voxel spacing of `0.0125` and half width `3.0`:

    dinovdb = OpenVDBLevelSet[dino, 0.0125, 3.0]

Screen Shot 2022-04-09 at 2 28 54 PM

See various properties:

    dinovdb["PropertyValueGrid"]

Screen Shot 2022-04-09 at 2 29 52 PM

### More involved example

Often times the voxel spacing, half width, and creator don't change in a workflow. There are optional global variables that can be set which let's the user not have to pass these values every time an object is created.

    $OpenVDBSpacing = 0.0125;
    $OpenVDBHalfWidth = 3.0;
    $OpenVDBCreator = "ghurst";

Let's recreate the dino vdb but this time omit the extra details:

    dinovdb = OpenVDBLevelSet[dino];

Next, let's drill a hole in our dino friend and apply 8 applications of a mean filter with window 2:

    OpenVDBDifferenceFrom[dinovdb, Cylinder[{{0, -2, 0.5}, {0, 2, 0.5}}, 0.5]];
    OpenVDBFilter[dinovdb, {"Mean", 2}, 8];

Now we can use the interactive viewer, which uses RayIntersector.h under the hood. The object can be rotated, dragged, and zoomed as discussed [here](https://reference.wolfram.com/language/workflow/RotatePanAndZoom3DGraphics.html).

    OpenVDBLevelSetViewer[dinovdb]

vdb

-------------

## TODO

* Some details I'd like to add before merging are documentation pages and message handling when user errors occur.
Example of desired documentation: https://reference.wolfram.com/language/ref/NDSolve.html
Example of desired messages:
Screen Shot 2022-04-09 at 3 16 22 PM

* The current list of supported functions are predominately skewed towards level sets at the moment. I'd like to port over more native vector and point functionality for example.
* More comments in the code.
* More information / design discussion in this PR. Right now I'm keeping this description short, but overtime more can be added if necessary.

------------

## Misc

* Mathematica's interrupter works with OpenVDB and is enabled with cmd-.
* OpenVDB can throw exceptions to Mathematica.
* Autocomplete where appropriate:

Screen Shot 2022-04-09 at 3 08 18 PM

* Highlighting to indicate missing / extra / invalid arguments:

Screen Shot 2022-04-09 at 3 10 33 PM

------------

## TODO as of Sept 26, 2022

* Verify ease of building on Windows and Linux
* README.md with build instructions, etc.
* Write a tutorial notebook chock full of examples
* Write Wolfram Community post highlighting this link
* _Stretch goal:_ free up front end during long VDB computations with LibraryLink's `createAsynchronousTaskWithThread`

------------

## TODO as of October 20, 2022

* Documentation touch ups
* Write Wolfram Community post highlighting this link (not blocking for merge & target for week of Oct 31 here)
* ~_Stretch goal:_ free up front end during long VDB computations with LibraryLink's `createAsynchronousTaskWithThread`~

------------

## TODO as of October 25, 2022

Ready to merge.

Companion PR for logo and documentation hosting: https://github.com/AcademySoftwareFoundation/openvdb-website/pull/66

* Write Wolfram Community post highlighting this link (not blocking for merge & target for week of Oct 31 here)
* ~_Stretch goal:_ free up front end during long VDB computations with LibraryLink's `createAsynchronousTaskWithThread`~

### CODE REVIEW

**Comment by Idclip** [comment]
+445k LOC :laughing: great work @ghurstunither and co

**Comment by ghurstunither** [comment]
> +445k LOC 😆 great work @ghurstunither and co

Thanks!

That number is a bit deceptive since the unit tests are stored in Wolfram notebooks (https://www.wolfram.com/notebooks/) -- which are not just plain text files. Those account for 396221 lines, leaving 49367 for the rest.

All the notebooks live in UnitTests/nb and look something like this in a text editor (they're meant to be open in Mathematica)

```
(* Content-type: application/vnd.wolfram.mathematica *)

(*** Wolfram Notebook File ***)
(* http://www.wolfram.com/nb *)

(* CreatedBy='Mathematica 13.0' *)

(*CacheID: 234*)
(* Internal cache information:
NotebookFileLineBreakTest
NotebookFileLineBreakTest
NotebookDataPosition[       158,          7]
NotebookDataLength[   1478964,      37624]
NotebookOptionsPosition[   1405194,      36149]
NotebookOutlinePosition[   1420949,      36475]
CellTagsIndexPosition[   1420906,      36472]
WindowTitle->Automatic
WindowFrame->Normal*)

(* Beginning of Notebook Content *)
Notebook[{

Cell[CellGroupData[{
Cell["CSG Tests", "Title",
 CellID->513597220,ExpressionUUID->"23e13f45-0548-42df-8805-52bb71eab504"],

Cell[CellGroupData[{

Cell["Generic", "Section",
 CellID->2114501238,ExpressionUUID->"f3e1e17c-182a-4ddb-96d0-59955511e922"],

Cell[CellGroupData[{

Cell["Initialization", "Subsection",
 CellID->806715902,ExpressionUUID->"deba74d8-32c0-4566-a3d3-9167de9eddec"],

Cell[CellGroupData[{
.
.
.
```

File: openvdb_wolfram/OpenVDBLink/Source/OpenVDBGrid/AggregateData.inl [hunk 1]
```diff
@@ -0,0 +1,204 @@
+#include "../Utilities/aggregate.h"
```
[the remark this answers is not in the document]
Comment (ghurstunither) [reply openvdb_wolfram/OpenVDBLink/Source/OpenVDBGrid/AggregateData.inl]: Thanks for pointing this out. It should be resolved now.

File: openvdb_wolfram/OpenVDBLink/Source/OpenVDBGrid.h [hunk 2]
```diff
@@ -0,0 +1,432 @@
...
+        return vec.rows() == n && vec.cols() == ValueLength;
+    }
+
+    inline wlGridPtr instanceGrid(mint id)
+    {
+        return mma::getInstance<OpenVDBGrid>(id).grid();
+    }
+};
+
+#include "OpenVDBGrid/AggregateData.inl"
```
[the remark this answers is not in the document]
Comment (ghurstunither) [reply openvdb_wolfram/OpenVDBLink/Source/OpenVDBGrid.h]: Thanks for pointing this out. It should be resolved now.

File: openvdb_wolfram/OpenVDBLink/Source/Utilities/distance.h [hunk 3]
```diff
@@ -0,0 +1,215 @@
+#pragma once
```
[the remark this answers is not in the document]
Comment (ghurstunither) [reply openvdb_wolfram/OpenVDBLink/Source/Utilities/distance.h]: Thanks for pointing this out. It should be resolved now.

File: openvdb_wolfram/OpenVDBLink/Source/OpenVDBGrid/AggregateData.h [hunk 4]
```diff
@@ -0,0 +1,207 @@
+// Copyright Contributors to the OpenVDB Project
+// SPDX-License-Identifier: MPL-2.0
+
+#include "../Utilities/Aggregate.h"
```
Comment (apradhana) [inline openvdb_wolfram/OpenVDBLink/Source/OpenVDBGrid/AggregateData.h:4]: Please consider including an include guard in this file. I see that the following files are missing include guards:
```
OpenVDBGrid/AggregateData.h
OpenVDBGrid/CSG.h
OpenVDBGrid/DistanceMeasure.h
OpenVDBGrid/Filter.h
OpenVDBGrid/FogVolume.h
OpenVDBGrid/IO.h
OpenVDBGrid/Image.h
OpenVDBGrid/LevelSetCreation.h
OpenVDBGrid/Measure.h
OpenVDBGrid/Mesh.h
OpenVDBGrid/Morphology.h
OpenVDBGrid/Render.h
OpenVDBGrid/Setters.h
OpenVDBGrid/Transform.h
```

File: openvdb_wolfram/OpenVDBLink/Source/OpenVDBGrid.h [hunk 5]
```diff
@@ -0,0 +1,438 @@
...
+#include <cmath>
+#include <ctime>
+#include <vector>
+
+using namespace openvdb;
+using namespace openvdb::tools;
+using namespace openvdbmma::types;
+using namespace openvdbmma::utils;
+
+
```
Comment (apradhana) [inline openvdb_wolfram/OpenVDBLink/Source/OpenVDBGrid.h:25]: Please consider wrapping this class in a namespace. The name `OpenVDBGrid` is generic enough that a user might use it, which may cause name collision if their library is linked against this sub-library.

File: openvdb_wolfram/OpenVDBLink/Source/OpenVDBGrid/AggregateData.h (hunk 4 above)
Comment (ghurstunither) [reply openvdb_wolfram/OpenVDBLink/Source/OpenVDBGrid/AggregateData.h:7]: This should be fixed now.

File: openvdb_wolfram/OpenVDBLink/Source/OpenVDBGrid.h (hunk 5 above)
Comment (ghurstunither) [reply openvdb_wolfram/OpenVDBLink/Source/OpenVDBGrid.h:25]: This has been added. It's now `openvdbmma::OpenVDBGrid`.

**Comment by kmuseth** [comment]
@ ghurstunither if you can point me to a free (try) version of Mathmatica then I  can try the installation now

**Comment by kmuseth** [comment]
> 

never mind - I found a free evaluation online

File: openvdb_wolfram/OpenVDBLink/Source/GlueTensors.h [hunk 6]
```diff
@@ -0,0 +1,420 @@
...
+
+#include <openvdb/openvdb.h>
+
+#include <tbb/parallel_for.h>
+#include <tbb/parallel_reduce.h>
+
+#include <vector>
+
+
+// ------------ Scalar classes ------------ //
```
Comment (apradhana) [inline openvdb_wolfram/OpenVDBLink/Source/GlueTensors.h:18]: Please consider enclosing this in a namespace.

Comment (ghurstunither) [reply openvdb_wolfram/OpenVDBLink/Source/GlueTensors.h:21]: This should be resolved now. They are now under the namespace `openvdbmma::glue`.

**Review by apradhana — APPROVED** [review]
I double checked the latest version and it builds fine on my Mathematica running on MacOS. Approving this.

**Review by kmuseth — CHANGES_REQUESTED** [review]
Currently the Mathematica extension is building using C++14, however OpenVDB v10 uses C++17. So I  believe this needs to be changed

**Comment by ghurstunither** [comment]
> Currently the Mathematica extension is building using C++14, however OpenVDB v10 uses C++17. So I believe this needs to be changed

Good catch, updated `BuildSettings.m` to use C++17.

**Review by kmuseth — APPROVED** [review]
I'm still not able to build and run this extension on my Mac, but it's likely due to the fact that I have local installations of OpenVDB and all its dependencies so in order not to block this PR from going into v10 I'll go ahead and approve it. Thanks for changing the compiler to C++17 as requested.

File: openvdb_wolfram/OpenVDBLink/Readme.md [hunk 7]
```diff
@@ -13,7 +13,9 @@ OpenVDB is an Academy Award-winning open-source C++ library comprising a novel h
 
 ### Documentation
 
-* [URL to docs?](.....)
+* [Web version](https://www.openvdb.org/documentation/wolfram)
```
Comment (apradhana) [inline openvdb_wolfram/OpenVDBLink/Readme.md:16]: https://www.openvdb.org/documentation/wolfram/index.html.

Comment (apradhana) [reply openvdb_wolfram/OpenVDBLink/Readme.md:16]: This is non-blocking.

**Pull request merged**